synapse/http/client.py

```diff
@@ -21,73 +21,218 @@
+            whitelist (netaddr.IPSet): The whitelisted IP addresses, that we can
+               request if it were otherwise caught in a blacklist.
+        """
         self.hs = hs
 
-        pool = HTTPConnectionPool(reactor)
-
-        # the pusher makes lots of concurrent SSL connections to sygnal, and
-        # tends to do so in batches, so we need to allow the pool to keep lots
-        # of idle connections around.
-        pool.maxPersistentPerHost = max((100 * CACHE_SIZE_FACTOR, 5))
-        pool.cachedConnectionTimeout = 2 * 60
-
-        # The default context factory in Twisted 14.0.0 (which we require) is
-        # BrowserLikePolicyForHTTPS which will do regular cert validation
-        # 'like a browser'
-        self.agent = Agent(
-            reactor,
-            connectTimeout=15,
-            contextFactory=hs.get_http_client_context_factory(),
-            pool=pool,
-        )
+        self._whitelist = whitelist
+        self._blacklist = blacklist
+        self._extra_treq_args = treq_args
+
         self.user_agent = hs.version_string
         self.clock = hs.get_clock()
+        self.reactor = hs.get_reactor()
         if hs.config.user_agent_suffix:
-            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix,)
+            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix)
 
         self.user_agent = self.user_agent.encode('ascii')
+        self._make_agent()
+
+    def _make_agent(self, _agent=False):
+
+        if _agent:
+            self.agent = _agent
+        else:
+            # the pusher makes lots of concurrent SSL connections to sygnal, and
+            # tends to do so in batches, so we need to allow the pool to keep
+            # lots of idle connections around.
+            pool = HTTPConnectionPool(self.reactor)
+            pool.maxPersistentPerHost = max((100 * CACHE_SIZE_FACTOR, 5))
+            pool.cachedConnectionTimeout = 2 * 60
+
+            # The default context factory in Twisted 14.0.0 (which we require) is
+            # BrowserLikePolicyForHTTPS which will do regular cert validation
+            # 'like a browser'
+            self.agent = Agent(
+                self.reactor,
+                connectTimeout=15,
+                contextFactory=self.hs.get_http_client_context_factory(),
+                pool=pool,
+            )
+
+        # If we have an IP blacklist, use the blacklisting Agent wrapper.
+        if self._blacklist:
+
+            nameResolver = IPBlacklistingResolver(
+                self.reactor, self._whitelist, self._blacklist
+            )
+
+            @implementer(IReactorPluggableNameResolver)
+            class Reactor(object):
+                def __getattr__(_self, attr):
+                    if attr == "nameResolver":
+                        return nameResolver
+                    else:
+                        return getattr(self.reactor, attr)
+
+            self.agent = Agent(
```

Comment:
suggestion: it'd be nice to avoid duplicating the Agent-creation code. How about setting `self.reactor = IPBlacklistingResolver(...)` in the constructor if there is a blacklist?

Reply:
done